### PR TITLE
elsi: improve package and add external libOMM

### PR DESCRIPTION
**Pull request by RMeli**
* Add MatrixSwitch package
* Add libOMM package
* Use libOMM as external in ELSI
  * Add include paths for all libraries
    * Fortran modules need one directory up (can't use `libs.directories` since they are just the `include/` directories)

### CODE REVIEW

**Comment by RMeli** [comment]
* Deprecate version `2.2.1` (it's quite old)
* Add `when=` to specific variants
* Use PEXSI as external with latest version (depends on SuperLU-Dist)

**Review by rasolca — APPROVED** [review]

**Comment by albestro** [comment]
@spackbot fix style

let's give it another chance...

File: var/spack/repos/builtin/packages/matrix-switch/package.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    variant("lapack", default=True, description="Build libOMM with LAPACK interface.")
+    variant("mpi", default=True, description="Build libOMM with MPI support.")
+    variant(
+        "scalapack",
+        default=True,
+        when="+mpi",
+        description="Build libOMM with ScaLAPACK interface.",
+    )
+    variant("dbcsr", default=False, when="+mpi", description="Build libOMM with DBCSR interface.")
```
Comment (albestro) [inline var/spack/repos/builtin/packages/matrix-switch/package.py]:
```suggestion
    variant("lapack", default=True, description="Build with LAPACK interface.")
    variant("mpi", default=True, description="Build with MPI support.")
    variant(
        "scalapack",
        default=True,
        when="+mpi",
        description="Build with ScaLAPACK interface.",
    )
    variant("dbcsr", default=False, when="+mpi", description="Build with DBCSR interface.")
```

File: var/spack/repos/builtin/packages/matrix-switch/package.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def cmake_args(self):
+        args = [
+            self.define_from_variant("WITH_LAPACK", "lapack"),
+            self.define_from_variant("WITH_MPI", "mpi"),
+            self.define_from_variant("WITH_SCALAPACK", "scalapack"),
+            self.define_from_variant("WITH_DBCSR", "dbcsr"),
+        ]
+
+        if self.spec.satisfies("+dbcsr"):
+            args.append(self.define("DBCSR_ROOT", self.spec["dbcsr"].prefix))
```
Comment (albestro) [inline var/spack/repos/builtin/packages/matrix-switch/package.py:47]: curiosity: why is this needed? cmake `find_package` looks in paths indicated by `*_ROOT` and then in `CMAKE_PREFIX_PATH` (which is set by spack in the build environment). Isn't it? Or is it there as a workaround for something specific?

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 3]
```diff
@@ -64,6 +86,13 @@ class Elsi(CMakePackage, CudaPackage):
     with when("+use_external_superlu"):
         depends_on("superlu-dist+cuda", when="+cuda")
         depends_on("superlu-dist~cuda", when="~cuda")
+    with when("+enable_pexsi +use_external_pexsi"):
+        depends_on("pexsi+fortran")
```
Comment (albestro) [inline var/spack/repos/builtin/packages/elsi/package.py:90]: Just checking: when `superlu-dist` is a `pexsi` dependency? that's why if `pexsi` is built internally we don't need to specify anything for it?

**Review by albestro — COMMENTED** [review]
Have a look at the style check output 😉

File: var/spack/repos/builtin/packages/elsi/package.py (hunk 3 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py:90]: Yes, `superlu-dist` is a `pexi` dependency and bundled internally: https://gitlab.com/elsi_project/elsi_interface/-/blob/master/CMakeLists.txt?ref_type=heads#L128-137

I'm declaring it explicitly here to propagate the `cuda` variant.

File: var/spack/repos/builtin/packages/matrix-switch/package.py (hunk 2 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/matrix-switch/package.py:47]: I think this stems from the issue that they are not using `find_package(DBCSR)` but they are using `find_package(CustomDbcsr)` and have a custom `FindCustomDbcsr.cmake`. I might be missing something, but it's the only way I managed to get it working.

**Comment by alalazo** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/matrix-switch/package.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class MatrixSwitch(CMakePackage):
+    """Intermediary interface between high-level routines for physics-related algorithms and low-level routines dealing
+    with matrix storage and manipulation."""
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/matrix-switch/package.py]:
```suggestion
    """Intermediary interface between high-level routines for 
    physics-related algorithms and low-level routines dealing
    with matrix storage and manipulation."""
```

**Comment by RMeli** [comment]
@spackbot fix style

**Review by albestro — APPROVED** [review]
LGTM
